File: javascript/041.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function(nums) {
  let i = 0;
  const n = nums.length;
  while(i < n) {
    if (nums[i] > n || nums[i] <= 0 || nums[i] === i + 1 || nums[i] === nums[nums[i] - 1]) {
      ++i;
    } else {
      const temp = nums[nums[i] - 1];
      nums[nums[i] - 1] = nums[i];
      nums[i] = temp;
    }
  }

  i = 0;
  while(i < n) {
    if (nums[i] !== i + 1) return i + 1;
    ++i;
  }

  return n + 1;
};

File: ruby/041.rb
# @param {Integer[]} nums
# @return {Integer}
def first_missing_positive(nums)
  length = nums.length
  0.upto(length - 1).each do |i|
    next if nums[i] > length || nums[i] <= 0
    while true do
      break if nums[i] > length || nums[i] <= 0
      target_index = nums[i] - 1
      break if target_index == i
      break if nums[i] == nums[target_index]
      nums[i], nums[target_index] = nums[target_index], nums[i]
    end
  end

  result = 1
  while nums[result - 1] == result do
    result += 1
  end
  
  result
end